### ISSUE TITLE
Bizarre circular type false positive

### ISSUE THREAD

**Issue by MichaelMitchell-at** [microsoft/TypeScript#48708]
# Bug Report

### 🔎 Search Terms

implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.(7022)

### 🕗 Version & Regression Information
This changed between versions 4.6.2 and 4.7.0-beta

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.7.0-dev.20220408#code/C4TwDgpgBA6gTgQzJOUC8UDeAoKepwQIAmA9gHYA2IUxCwCAXFOQK4C2ARhKgD4utKlANzYAvqOwAzVuQDGwAJYUoEAB4J2YShAAUAd0TIezeEhQBKLLnyKpUA0ZQA6Og3RoMbIVcLBWcOSS+FBSpKi6chQAzsBQAPpQpPYA2gC6VjghIVHksbToUIbmPK70CKLZ+Ln5cRjElVVQNXGUivkYKQAMaY1Vdg5tsSnAGVhiNngTYkA)

### 💻 Code

```ts repro
type Wrapper = {
    data: number | null;
};

function example(wrapper: Wrapper) {
    // The error goes away if this line is removed and `null` is removed from the type
    if (wrapper.data === null) return;

    // The error goes away if the loop body is inlined o_O
    for (const _ of []) {
        const d = wrapper.data;
        //    ^ 'd' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.

        // The error goes away without these lines o_O
        const t = d;
        const list = [0];
        if (list[t]) {}
    }
}
```

### 🙁 Actual behavior
Error: implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer

### 🙂 Expected behavior
No error

**Comment by MichaelMitchell-at** [comment]
cc @ahejlsberg since you made some significant type flow analysis changes

**Comment by softwareCobbler** [comment]
Have a similar repro; was looking to file a bug and this seems like the right place for it.

```
interface A {
    m: string
}

function foo(a: A) : void {
    let working : A | undefined = a;

    while (working) {
        // 'xxx' has type 'any' because it is referenced in its own initializer
        const xxx = working.m;
        if (xxx) {
            working = xxx;
        }
    }
}

```

[playground link](https://www.typescriptlang.org/play?ts=4.7.2#code/JYOwLgpgTgZghgYwgAgILIN4FgBQz-IC2AXMgM5hSgDmuAvrrjAK4gJjAD2IyMnnACjilUASmSkAbp2AATTLgLIANhDDIA7pygBrGhLTIAPslayIMUBHkBeZHADcjPAQ0ALYKuQCtumuOwXJXwAehDkAHIADxiI5Dc4MmQwAE8ABxQIuBAUuIAjCAQ4ZjIUYHVgJKgLaAg2a2RQRrAkzg0eUHLgOGVgAC9oRWD8BG4KZBio5DtfPRBqADpCJyDg4BhvSYCh4aVZ-TtJld3kBlWzs6A)

**Comment by MichaelMitchell-at** [comment]
I looked a bit into what part of https://github.com/microsoft/TypeScript/commit/162713fac9da0304769ef0b7cfdd568f319a55ea caused this regression. It seems like adding `isEntityNameExpression(expr.argumentExpression)` to `isNarrowableReference` allows `flowNode`s to be set on expressions where they weren't previously during the binding phase. This in turn seems to lead to a circular type resolution when typing the expression along the `getFlowTypeOfReference` code path.

I haven't looked deep enough yet to understand why the circular type resolution happens yet, but I'm sharing my findings here in case it helps someone else pick up this investigation.

**Comment by MichaelMitchell-at** [comment]
@Andarist you've been pretty capable at fixing random bugs I've stumbled upon before, so maybe this one would be of interest to you 😉

**Comment by vassudanagunta** [comment]
The error also goes away if you simple move just the initialization out of the loop:

```ts
type Wrapper = {
    readonly data: number | null;
};

function example(wrapper: Wrapper) {
    if (wrapper.data === null) return;

    const d = wrapper.data;
    for (const _ of []) {
        const t = d;
        const list = [0];
        if (list[t]) {}
    }
}
```


Also, per the Playground link, the change happened after 4.6.4. I presume https://github.com/microsoft/TypeScript/commit/162713fac9da0304769ef0b7cfdd568f319a55ea is post 4.6.4?
